Add shellcheck to pre-commit hook for instant local feedback

The hooks/pre-commit currently only validates YAML schemas. Running shellcheck locally before push would catch issues before CI, matching the same checks in lint.yml. Add a shellcheck invocation at the top of hooks/pre-commit (and update the 'just install-hooks' recipe if needed).

_Follow-up from #45_

# Implementation Plan

Now I have a complete picture. Here's the implementation plan:

---

## Implementation Plan: Issue #215 — Add shellcheck to pre-commit hook

### 1. Objective

Add a `shellcheck` invocation to `hooks/pre-commit` so contributors get instant local feedback on shell script quality before pushing. The check should mirror what CI (`lint.yml`) runs, covering `scripts/` and `hooks/` shell files. Also add a `.shellcheckrc` at repo root (currently missing — only exists in a stale worktree) and update the `just install-hooks` recipe to report on the shellcheck dependency.

---

### 2. Approach

- **Prepend** the shellcheck check to `hooks/pre-commit` so it runs before YAML validation (fast fail on the most common change type — shell scripts — without requiring `check-jsonschema` or `yq`).
- **Scope to staged shell files** using `git diff --cached --name-only` filtered to `scripts/` and `hooks/` with a shell extension filter. This avoids running shellcheck on the whole repo on every commit.
- **Graceful degradation**: if `shellcheck` is not installed, print a helpful install message and exit 1 (same pattern as the existing `check-jsonschema` and `yq` guards). Do **not** silently skip — that defeats the purpose.
- **No `.pre-commit-config.yaml`**: the repo uses a hand-crafted `hooks/pre-commit`, not the `pre-commit` framework. Keep that pattern; add shellcheck inline.
- **Add `.shellcheckrc`** at repo root (copy/adapt the one found in the stale worktree) as the single source of truth for suppressions.
- **Update `just install-hooks`** to also print whether `shellcheck` is available, so the developer knows their hook is fully armed.

---

### 3. Files to Create

| File | Description |
|------|-------------|
| `.shellcheckrc` | Repo-root shellcheck config. Sets `severity=warning`, `shell=bash`, disables `SC1091` (sourced files). Adapted from the stale worktree version. |

---

### 4. Files to Modify

#### `hooks/pre-commit`

**Add** a new section immediately after the `set -euo pipefail` block and before the YAML validation logic:

```bash
# --- ShellCheck: staged shell scripts ---
STAGED_SHELL=$(git diff --cached --name-only --diff-filter=ACM \
  | grep -E '^(scripts|hooks)/.*\.(sh|bash)$' || true)
# Also catch extensionless shell files by checking the shebang
STAGED_EXTENSIONLESS=$(git diff --cached --name-only --diff-filter=ACM \
  | grep -E '^(scripts|hooks)/[^/]+$' | grep -v '\.' || true)

ALL_SHELL_FILES=""
for f in $STAGED_SHELL $STAGED_EXTENSIONLESS; do
  if head -1 "${REPO_ROOT}/${f}" 2>/dev/null | grep -qE '^#!.*\b(bash|sh)\b'; then
    ALL_SHELL_FILES="${ALL_SHELL_FILES} ${REPO_ROOT}/${f}"
  fi
done
# Always include .sh/.bash files without shebang check
for f in $STAGED_SHELL; do
  ALL_SHELL_FILES="${ALL_SHELL_FILES} ${REPO_ROOT}/${f}"
done
# Deduplicate
ALL_SHELL_FILES=$(echo "$ALL_SHELL_FILES" | tr ' ' '\n' | sort -u | tr '\n' ' ')

if [[ -n "$ALL_SHELL_FILES" ]]; then
  if ! command -v shellcheck &>/dev/null; then
    echo "ERROR: shellcheck not found — cannot lint shell scripts before commit" >&2
    echo "  Install: https://github.com/koalaman/shellcheck#installing" >&2
    exit 1
  fi
  echo "Running shellcheck on staged shell files..."
  # shellcheck disable=SC2086  # word-splitting on ALL_SHELL_FILES is intentional
  if ! shellcheck $ALL_SHELL_FILES; then
    echo ""
    echo "Pre-commit: shellcheck failed. Fix the issues above before committing."
    exit 1
  fi
  echo "shellcheck: all staged shell files OK."
fi
```

> **Key implementation note from team knowledge**: `hooks/pre-commit` does not use the `pre-commit` framework, so the `bash -c` pitfall (losing filenames as `$0`) does not apply here. `shellcheck` is called directly with expanded filenames.

**Simplify the actual implementation** — the staged file detection above can be tightened. See Implementation Order step 2 for the clean final form.

---

#### `justfile`

Update the `install-hooks` recipe to check for `shellcheck` availability:

```justfile
# Install the pre-commit hook into .git/hooks/
install-hooks:
    cp hooks/pre-commit .git/hooks/pre-commit
    chmod +x .git/hooks/pre-commit
    @echo "pre-commit hook installed."
    @if command -v shellcheck > /dev/null 2>&1; then \
        echo "  shellcheck: found ($(shellcheck --version | head -2 | tail -1))"; \
    else \
        echo "  shellcheck: NOT FOUND — hook will fail on shell file commits"; \
        echo "  Install: https://github.com/koalaman/shellcheck#installing"; \
    fi
```

---

### 5. Implementation Order

1. **Create `.shellcheckrc`** at repo root — baseline suppressions (`severity=warning`, `shell=bash`, `disable=SC1091`). This must exist before running shellcheck so the hook and CI use the same configuration.

2. **Edit `hooks/pre-commit`** — insert the shellcheck block after the `set -euo pipefail` / variable setup block, before the `STAGED_YAMLS` block. Specifically:
   - After line establishing `REPO_ROOT`
   - Detect staged files in `scripts/` and `hooks/` matching `*.sh`, `*.bash`, or extensionless with a bash/sh shebang
   - Guard on `shellcheck` availability with a clear error
   - Run `shellcheck` on the collected files
   - Exit 1 with a clear message on failure

3. **Edit `justfile`** — update the `install-hooks` recipe to print shellcheck availability status.

4. **Verify** (see section 6).

---

### 6. Verification

```bash
# 1. Confirm .shellcheckrc is picked up
shellcheck --list-optional   # should not error

# 2. Stage a shell file with a known shellcheck violation, try to commit
echo 'echo $UNQUOTED' >> scripts/plan.sh   # SC2086 — but severity=warning so it fires
git add scripts/plan.sh
git commit -m "test"   # should be blocked by hook

# 3. Stage a clean shell file — commit should succeed
git restore scripts/plan.sh

# 4. Stage only a YAML file — shellcheck section should be skipped entirely (no "shellcheck not found" error)

# 5. Run just install-hooks and confirm shellcheck status line appears

# 6. Confirm existing YAML validation still works end-to-end
git add agents/hermes/someagent.yaml
git commit -m "test yaml"
```

---

### 7. Skills Used

**Invoked during planning:**
- `hephaestus:advise` (via Prior Learnings injection) — retrieved team knowledge on ShellCheck configuration and pre-commit hook patterns

**Team knowledge base skills referenced in Prior Learnings:**
- `shellcheck-scope-templates` — ShellCheck configuration, `.shellcheckrc` scoping, pre-commit hook integration
- `pre-commit-hook-configuration` — adding/configuring local pre-commit hooks, `pass_filenames`, `language: system`, hook stanza patterns
- `ci-cd-failure-fix-patterns` — `bash -c` arg-passing pitfall (Pattern 3); confirmed this does **not** apply here since `hooks/pre-commit` calls `shellcheck` directly, not via `bash -c`

---
*Generated by Claude Code Planner*

--- a/.github/workflows/validate.yml
+++ b/.github/workflows/validate.yml
@@ -7,12 +7,13 @@ on:
       - 'agents/**'
       - 'fleets/**'
       - 'scripts/**'
+      - 'hooks/**'
       - '.github/workflows/**'
       - 'schemas/**'
       - '.myrmidons.yaml'
+      - '.pre-commit-config.yaml'
       - 'pixi.toml'
       - 'pixi.lock'
-      - 'hooks/**'
       - 'tests/**'
       - 'CHANGELOG.md'
 
@@ -125,6 +126,19 @@ jobs:
           echo "To skip this check, include [skip changelog] in the PR title, body,"
           echo "any commit message, or add the 'skip changelog' label to the PR."
           exit 1
+
+  pre-commit:
+    name: Pre-commit hooks
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+
+      - uses: ./.github/actions/setup-pixi
+        with:
+          environment: lint
+
+      - name: Run pre-commit on all files
+        run: pixi run --environment lint pre-commit run --all-files
 
   validate:
     name: Validate YAML Schemas

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -1,7 +1,7 @@
 repos:
   # Standard pre-commit hooks
   - repo: https://github.com/pre-commit/pre-commit-hooks
-    rev: v5.0.0
+    rev: cef0300fd0fc4d2a87a85fa2093c6b283ea36f4b  # v5.0.0
     hooks:
       - id: trailing-whitespace
       - id: end-of-file-fixer
@@ -10,14 +10,14 @@ repos:
 
   # Shell script linting
   - repo: https://github.com/shellcheck-py/shellcheck-py
-    rev: v0.10.0.1
+    rev: a23f6b85d0fdd5bb9d564e2579e678033debbdff  # v0.10.0.1
     hooks:
       - id: shellcheck
         args: [--severity=warning]
 
   # YAML linting for agents/ and fleets/
   - repo: https://github.com/adrienverge/yamllint
-    rev: v1.35.1
+    rev: 81e9f98ffd059efe8aa9c1b1a42e5cce61b640c6  # v1.35.1
     hooks:
       - id: yamllint
         files: ^(agents|fleets)/.*\.yaml$

--- a/pixi.toml
+++ b/pixi.toml
@@ -13,7 +13,7 @@ bats-core = ">=1.11,<2"
 curl      = ">=8.0"
 
 [feature.lint.dependencies]
-shellcheck = ">=0.10,<1"
+shellcheck = "0.10.0.*"
 yamllint   = ">=1.35,<2"
 pre-commit = ">=3.0,<5"
 

